tests/ubench/event_setup.cpp

```diff
@@ -0,0 +1,207 @@
+    std::vector<size_t> part(ncells+1);
+
+    while (state.KeepRunning()) {
+        part[0] = 0;
+
+        // push events into the queue corresponding to target cell
+        for (const auto& e: input_events) {
+            event_lanes[e.target.gid].push(e);
+        }
+
+        // pop from queue to form single sorted vector
+        std::vector<pev> staged_events;
+        staged_events.reserve(input_events.size());
+        size_t i=0;
+        for (auto& lane: event_lanes) {
+            while (auto e = lane.pop_if_before(1.f)) {
+                staged_events.push_back(*e);
+            }
+            part[++i] = staged_events.size();
+        }
+
+        benchmark::ClobberMemory();
+    }
+}
+
+void n_vector(benchmark::State& state) {
+    using pev = postsynaptic_spike_event;
+    const std::size_t ncells = state.range(0);
+    const std::size_t ev_per_cell = state.range(1);
+
+    auto input_events = generate_inputs(ncells, ev_per_cell);
+
+    // state
+    std::vector<std::vector<pev>> event_lanes(ncells);
+    std::vector<size_t> part(ncells+1);
+    std::vector<decltype(input_events.begin())> ext(ncells);
+
+    struct ev_lt_pred {
+        bool operator()(float t, const pev& ev) { return t<ev.time; }
+        bool operator()(const pev& ev, float t) { return ev.time<t; }
+    };
+
+    // NOTE: this is a "full" implementation, that can handle the case where
+    // input_events contains events that are to be delivered after the current
+    // delivery interval. The event_lanes vectors keep undelivered events.
+    while (state.KeepRunning()) {
+        ext.clear();
+
+        // push events into a per-cell vectors (unsorted)
+        for (const auto& e: input_events) {
+            event_lanes[e.target.gid].push_back(e);
+        }
+        // sort each per-cell queue and keep track of the subset of sorted
+        // events that are to be delivered in this interval.
+        for (auto& lane: event_lanes) {
+            std::sort(lane.begin(), lane.end(),
+                      [](const pev& l, const pev& r) {return l.time<r.time;});
+            ext.push_back(
+                std::lower_bound(lane.begin(), lane.end(), 1.f, ev_lt_pred()));
+        }
+        // calculate partition of output buffer according to target cell gid
+        part[0] = 0;
+        for (size_t i=0; i<ncells; ++i) {
+            part[i+1] = part[i] + std::distance(event_lanes[i].begin(), ext[i]);
+        }
+        // copy events into the output flat buffer
+        std::vector<postsynaptic_spike_event> staged_events(part.back());
+        auto b = staged_events.begin();
+        for (size_t i=0; i<ncells; ++i) {
+            std::copy(event_lanes[i].begin(), ext[i], b+part[i]);
+        }
+
+        // remove events that were delivered from the event lanes
+        auto i=0u;
+        for (auto& lane: event_lanes) {
+            lane.erase(lane.begin(), ext[i++]);
+        }
+
+        benchmark::ClobberMemory();
+    }
```

Comment:
For consistency across runs, if the generated event set is extended to include events after time `1.f`, then the lanes will grow in each iteration of the benchmark loop; they should be cleared within the loop.